src/quickfix.c

```diff
@@ -1958,6 +1965,34 @@ qf_get_curlist(qf_info_T *qi)
     return qf_get_list(qi, qi->qf_curlist);
 }
 
+/*
+ * Pop a quickfix list from the quickfix/location list stack
+ * Automatically adjust qf_curlist so that it stays pointed
+ * to the same list, unless it is deleted, if so then use the
+ * newest created list instead. qf_listcount will be set correctly.
+ * The above will only happen if <adjust> is TRUE.
+ */
+    static void
+qf_pop_stack(qf_info_T *qi, int adjust)
+{
+    int i;
+    qf_free(&qi->qf_lists[0]);
+    for (i = 1; i < qi->qf_listcount; ++i)
+	qi->qf_lists[i - 1] = qi->qf_lists[i];
+
+    // fill with zeroes now unused list at the top
+    vim_memset(qi->qf_lists + qi->qf_listcount - 1, 0, sizeof(*qi->qf_lists));
+
+    if (adjust)
+    {
+	qi->qf_listcount--;
+	if (qi->qf_curlist == 0)
+	    qi->qf_curlist = qi->qf_listcount - 1;
+	else
```

Comment:
If qi->qf_curlist is 0, then there is no need to do anything, right?

```suggestion
	if (qi->qf_curlist > 0)
```

Reply:
I mean the behaviour is that when the current list is removed from the stack, the new current list should be the most recent one (the list at top, so qf_listcount - 1)

Reply:
Hmm...
[0] is the oldest list, and when it is deleted, shouldn't it just remain pointing to the (new) oldest list ([0])?

Reply:
I was wondering about that but I decided to stick with what @chrisbra [said](https://github.com/vim/vim/pull/16920#discussion_r2004207565)

Reply:
Find `CHECK-THIS` in this comment.

---

## Action 1.
### Terminal #1

```bash
$ cd /path-to-your-master-branch-vim/src
$ vim --clean
```

```vim
:vimgrep eval0 *.c
:chistory
> error list 1 of 1; 23 errors    :vimgrep eval0 *.c
Press ENTER or type command to continue
```

### Terminal #2

```bash
$ cd /path-to-your-master-branch-vim/src
$ gdb ./vim $(pgrep -n vim)
```
```gdb
(gdb) p ql_info.qf_listcount
$1 = 1
(gdb) p ql_info.qf_curlist
$2 = 0
(gdb) p ql_info.qf_lists[0].qf_title
$3 = (char_u *) 0x570e8820e4f0 ":vimgrep eval0 *.c"
(gdb) q
A debugging session is active.

        Inferior 1 [process 4314] will be detached.

Quit anyway? (y or n) y
```

---

## Action 2.
### Terminal #1

```vim
:vimgrep eval1 *.c
:chistory
  error list 1 of 2; 23 errors    :vimgrep eval0 *.c
> error list 2 of 2; 34 errors    :vimgrep eval1 *.c
Press ENTER or type command to continue
```

### Terminal #2

```bash
$ gdb ./vim $(pgrep -n vim)
```
```gdb
(gdb) p ql_info.qf_listcount
$1 = 2
(gdb) p ql_info.qf_curlist
$2 = 1
(gdb) p ql_info.qf_lists[0].qf_title
$3 = (char_u *) 0x570e8820e4f0 ":vimgrep eval0 *.c"
(gdb) p ql_info.qf_lists[1].qf_title
$4 = (char_u *) 0x570e8860ff60 ":vimgrep eval1 *.c"
(gdb) q
A debugging session is active.

        Inferior 1 [process 4314] will be detached.

Quit anyway? (y or n) y
```

---

## Action 3.
### Terminal #1

```vim
:colder
:chistory
> error list 1 of 2; 23 errors    :vimgrep eval0 *.c
  error list 2 of 2; 34 errors    :vimgrep eval1 *.c
Press ENTER or type command to continue
```

### Terminal #2

```bash
$ gdb ./vim $(pgrep -n vim)
```
```gdb
(gdb) p ql_info.qf_listcount
$1 = 2
(gdb) p ql_info.qf_curlist
$2 = 0		// CHECK-THIS `:colder` changed it from 1 to 0.
(gdb) p ql_info.qf_lists[0].qf_title
$3 = (char_u *) 0x570e8820e4f0 ":vimgrep eval0 *.c"
(gdb) p ql_info.qf_lists[1].qf_title
$4 = (char_u *) 0x570e8860ff60 ":vimgrep eval1 *.c"
(gdb) q
A debugging session is active.

        Inferior 1 [process 4314] will be detached.

Quit anyway? (y or n) y
```

Reply:
Isn't that expected behaviour? Since element 0 in the stack is the older one.